core/src/main/java/io/opentelemetry/android/SdkPreconfiguredRumBuilder.kt

```diff
@@ -50,6 +51,14 @@ class SdkPreconfiguredRumBuilder
             // might turn off/on additional telemetry depending on whether the app is active or not
             appLifecycleService.registerListener(timeoutHandler)
 
+            val eventLogger =
+                SdkEventLoggerProvider.create(sdk.logsBridge)
+                    .get(OpenTelemetryRum::class.java.simpleName)
+
+            sessionManager.addObserver(SessionIdEventSender(eventLogger))
+            // After addObserver(), we call getSessionId() to trigger a session.start event
```

Comment:
This seems a bit forced. I'm guessing it will happen naturally when the first telemetry event is recorded, is there a reason why we must start the session earlier?

Reply:
probably needed if we want a session to be active as early as possible.
eg SDK is initiated during the Application#onCreate class.
If the session starts (automatically) when the OS gets a notification that the app is in the foreground (using androidx.lifecycle), that might be too late because the whole Application#onCreate code has been run, maybe some telemetry data was collected, but there was not an active session yet, in this case, makes sense that the SDK starts as session ASAP.

Reply:
Also, if we must do it earlier than the first telemetry event, which I guess would depend on the definition of a session, for example, if the definition ends up telling something around the lines of "the session starts when the app is launched", then I think the initialization of the SDK is too late to do so as well. It reminds me of the app launch timer that was recently mentioned [here](https://github.com/open-telemetry/opentelemetry-android/issues/560), where we might have to come up with some other mechanisms to track the start time temporarily until the Otel instance is ready.

Reply:
> maybe some telemetry data was collected, but there was not an active session yet

The session is queried and applied in the processors, so it will start as soon as the first telemetry data is collected. We do have an exception though where we collect events before the SDK is finished initializing which is [here](https://github.com/open-telemetry/opentelemetry-android/blob/5a230fc59e85b244ff253c8faabe023c353c9bd4/instrumentation/startup/src/main/java/io/opentelemetry/android/instrumentation/startup/SdkInitializationEvents.kt#L22) in the startup instrumentation where we store events in memory until those can be sent after the SDK is ready. It's probably worth considering a similar solution for this case in case we really need to trigger the event before the first telemetry data is collected.

Reply:
I too think the first session should start as early as possible, but if the OTel SDK isn't started, can we even do anything with it? If we simply initialize it lazily when an instance of the processor is created, it should ensure that every piece of OTel telemetry recorded will have a session ID?

Reply:
> If we simply initialize it lazily when an instance of the processor is created, it should ensure that every piece of OTel telemetry recorded will have a session ID?

The session is started when `getSessionId()` is called for the first time. It happens as soon as the [first span is started](https://github.com/open-telemetry/opentelemetry-android/blob/56dbb88b5e07235eb14c30ea521788070b81fe19/core/src/main/java/io/opentelemetry/android/SessionIdSpanAppender.java#L26) or the [first log is emitted](https://github.com/open-telemetry/opentelemetry-android/blob/56dbb88b5e07235eb14c30ea521788070b81fe19/core/src/main/java/io/opentelemetry/android/SessionIdLogRecordAppender.kt#L20), so there's no way that a signal is created without a session id. So it seems to me that we don't need to force it unless we must create a session before any telemetry is sent, like when the app is created for example, in which case this wouldn't be a good place to do so either.

I'd prefer to avoid adding behaviors to `core` unless it's a must for everyone, so it seems to me that the agent would be a nice place for this kind of behavior. I'm just back from PTO so I was planning to continue the work on [this prototype](https://github.com/open-telemetry/opentelemetry-android/pull/624) as it seems like the overall idea has gotten thumbs-up, probably during the same work or right after I can add this session event sender behavior to the agent, wdyt @breedx-splk ?